File: Anul II/Medii de proiectare si programare/Laborator/lab07/Ubb.BikeContest.Repository/src/TeamDbRepository.cs
using System.Collections;
using System.Data;
using log4net;
using Ubb.BikeContest.Model;

namespace Ubb.BikeContest.Repository;

public class TeamDbRepository : ITeamRepository
{
    private static readonly ILog Log = LogManager.GetLogger("TeamDbRepository");
    private readonly IDictionary<string, string?> _props;

    public TeamDbRepository(IDictionary<string, string?> props)
    {
        Log.Info("Creating TeamDbRepository ");
        _props = props;
    }

    public Team FindById(long id)
    {
        Log.InfoFormat("Entering Get with value {0}", id);
        IDbConnection connection = DbUtils.GetConnection(_props);

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "SELECT id,name FROM teams WHERE id=@id";
            IDbDataParameter paramId = command.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = id;
            command.Parameters.Add(paramId);

            using (var dataReader = command.ExecuteReader())
            {
                if (dataReader.Read())
                {
                    Team team = Extract(dataReader);
                    Log.InfoFormat("Exiting Get with value {0}", team);
                    return team;
                }
            }
        }
        Log.InfoFormat("Exiting Get with value {0}", null);
        return null;
    }

    public IEnumerable<Team> FindAll()
    {
        Log.InfoFormat("Entering Read");
        IDbConnection connection = DbUtils.GetConnection(_props);
        IList<Team> teams = new List<Team>();

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "SELECT id,name FROM teams";
            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    Team team = Extract(dataReader);
                    teams.Add(team);
                }
            }
        }
        Log.InfoFormat("Exiting Read with value {0}", teams);
        return teams;
    }

    public void Save(Team newEntity)
    {
        Log.InfoFormat("Entering Add with value {0}", newEntity);
        IDbConnection connection = DbUtils.GetConnection(_props);

        using (var command = connection.CreateCommand())
        {
            command.CommandText = 
                "INSERT INTO teams(name) VALUES(@name);";
            var name = command.CreateParameter();
            name.ParameterName = "@name";
            name.Value = newEntity.Name;
            command.Parameters.Add(name);

            var result = command.ExecuteNonQuery();
            Log.InfoFormat("Added {0} entities", result);
        }
    }

    public void Delete(long id)
    {
        Log.InfoFormat("Entering Delete with value {0}", id);
        IDbConnection connection = DbUtils.GetConnection(_props);
        using (var command = connection.CreateCommand())
        {
            command.CommandText = "DELETE FROM teams WHERE id=@id";
            var paramId = command.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = id;
            command.Parameters.Add(paramId);
            var result = command.ExecuteNonQuery();
            Log.InfoFormat("Deleted {0} entities", result);
        }
    }

    public void Update(Team updatedEntity)
    {
        throw new NotImplementedException();
    }

    public Team GetTeamByName(string teamName)
    {
        Log.InfoFormat("Entering GetTeamByName with value {0}", teamName);
        var connection = DbUtils.GetConnection(_props);

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "SELECT id,name FROM teams WHERE name=@name";
            var name = command.CreateParameter();
            name.ParameterName = "@id";
            name.Value = teamName;
            command.Parameters.Add(name);

            using (var dataReader = command.ExecuteReader())
            {
                if (dataReader.Read())
                {
                    Team team = Extract(dataReader);
                    Log.InfoFormat("Exiting GetTeamByName with value {0}", team);
                    return team;
                }
            }
        }
        Log.InfoFormat("Exiting GetTeamByName with value {0}", null);
        return null;
    }
    
    private static Team Extract(IDataReader dataReader)
    {
        var id = dataReader.GetInt64(0);
        var name = dataReader.GetString(1);

        var team = new Team(name)
        {
            Id = 0
        };
        team.Id = id;
        return team;
    }
}

File: Anul II/Medii de proiectare si programare/Laborator/lab09/persistence/src/main/java/com/ubb/repository/jdbc/TeamDBRepository.java
package com.ubb.repository.jdbc;

import com.ubb.model.Team;
import com.ubb.repository.ITeamRepository;
import com.ubb.utils.JDBCUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Optional;
import java.util.Properties;

public class TeamDBRepository implements ITeamRepository {
    private final JDBCUtils jdbcUtils;
    private static final Logger logger = LogManager.getLogger();

    public TeamDBRepository(Properties properties) {
        logger.info("Initializing TeamDBRepository with properties: {} ", properties);
        jdbcUtils = new JDBCUtils(properties);
    }

    @Override
    public Optional<Team> getTeamByName(String teamName) {
        logger.traceEntry("getTeamByName with task {} ", teamName);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "SELECT * FROM teams WHERE name=?")) {
            preparedStatement.setString(1, teamName);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                if (resultSet.next()) {
                    Team team = extract(resultSet);
                    logger.traceExit(team);
                    return Optional.of(team);
                }
            }
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
        logger.traceExit("null");
        return Optional.empty();
    }

    @Override
    public Optional<Team> findById(Long id) {
        logger.traceEntry();
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "SELECT * FROM teams WHERE id=?")) {
            preparedStatement.setLong(1, id);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                if (resultSet.next()) {
                    Team team = extract(resultSet);
                    logger.traceExit(team);
                    return Optional.of(team);
                }
            }
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
        logger.traceExit("null");
        return Optional.empty();
    }

    @Override
    public Collection<Team> findAll() {
        logger.traceEntry();
        Connection connection = jdbcUtils.getConnection();
        Collection<Team> teams = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "SELECT * FROM teams")) {
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Team team = extract(resultSet);
                    teams.add(team);
                }
            }
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
        logger.traceExit(teams);
        return teams;
    }

    @Override
    public void save(Team newEntity) {
        logger.traceEntry("saving task {} ", newEntity);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "INSERT INTO teams (name) VALUES (?)")) {
            preparedStatement.setString(1, newEntity.getName());
            int result = preparedStatement.executeUpdate();
            logger.traceExit("Saved {} instances");
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
    }

    @Override
    public void delete(Long id) {
        logger.traceEntry("deleting task of id={} ", id);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "DELETE FROM teams WHERE id=?")) {
            preparedStatement.setLong(1, id);
            int result = preparedStatement.executeUpdate();
            logger.traceExit("Deleted {} instances", result);
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
    }

    @Override
    public void update(Team updatedEntity) {
    }

    private Team extract(ResultSet resultSet) throws SQLException {
        Long id = resultSet.getLong("id");
        String name = resultSet.getString("name");

        Team team = new Team(name);
        team.setId(id);
        return team;
    }
}
